src/probnum/filtsmooth/particlefiltsmooth/_particle_filter.py

```diff
@@ -0,0 +1,218 @@
+        # Any linearised measurement model that could be used in a
+        # Gaussian filter can be used here and will likely be a better
+        # choice than the bootstrap.
+        self.linearized_measurement_model = linearized_measurement_model
+
+    def filter(self, dataset, times):
+        """Apply Particle filtering to a data set.
+
+        Parameters
+        ----------
+        dataset : array_like, shape (N, M)
+            Data set that is filtered.
+        times : array_like, shape (N,)
+            Temporal locations of the data points.
+            The zeroth element in times and dataset is the location of the initial random variable.
+
+        Returns
+        -------
+        ParticleFilterPosterior
+            Posterior distribution of the filtered output.
+        """
+
+        # Initialize: condition on first data point using the initial random
+        # variable as a dynamics rv (i.e. as the current prior).
+        particles_and_weights = np.array(
+            [
+                self.compute_new_particle(dataset[0], times[0], self.initrv)
+                for _ in range(self.num_particles)
+            ],
+            dtype=object,
+        )
+        particles = np.stack(particles_and_weights[:, 0], axis=0)
+        weights = np.stack(particles_and_weights[:, 1], axis=0)
+        weights = np.array(weights) / np.sum(weights)
+        curr_rv = random_variables.Categorical(support=particles, probabilities=weights)
+        rvs = [curr_rv]
+
+        for idx in range(1, len(times)):
+            curr_rv, _ = self.filter_step(
+                start=times[idx - 1],
+                stop=times[idx],
+                randvar=curr_rv,
+                data=dataset[idx],
+            )
+            rvs.append(curr_rv)
+        return ParticleFilterPosterior(rvs, times)
+
+    def filter_step(self, start, stop, randvar, data):
+        """Perform a particle filter step.
+
+        It consists of the following steps:
+        1. Propagating the "past" particles through the dynamics model.
+        2. Computing a "proposal" random variable.
+        This is either the prior dynamics model or the output of a filter step
+        of an (approximate) Gaussian filter.
+        3. Sample from the proposal random variable. This is the "new" particle.
+        4. Propagate the particle through the measurement model.
+        This is required in order to require the PDF of the resulting RV at
+        the data. If this is small, the weight of the particle will be small.
+        5. Compute weights ("event probabilities") of the new particle.
+        This requires evaluating the PDFs of all three RVs (dynamics, proposal, measurement).
+
+        After this is done for all particles, the weights are normalized in order to sum to 1.
+        """
+        new_weights = randvar.probabilities.copy()
+        new_support = randvar.support.copy()
+
+        for idx, (particle, weight) in enumerate(zip(new_support, new_weights)):
```

Comment:
these loops are slow, but at least they are rather easy to read (and I had no idea how to optimize them so that the difference is more than incremental)